prusti-contracts/prusti-contracts/src/core_spec/mem.rs

```diff
@@ -0,0 +1,65 @@
+use crate::*;
+
+#[extern_spec]
+mod core {
+    mod mem {
+        use crate::*;
+        
+        #[pure]
+        #[ghost_constraint(T: core_spec::mem::KnownSize, [ensures(result == T::size())])]
+        fn size_of<T>() -> usize;
+        
+        #[pure]
+        #[ghost_constraint(T: core_spec::mem::KnownSize, [ensures(result == T::align())])]
+        fn align_of<T>() -> usize;
+		
+		#[ensures(*x === old(snap(y)) && *y === old(snap(x)))]
+		fn swap<T>(x: &mut T, y: &mut T);
+    }
+}
+
+pub trait KnownSize {
+    #[pure]
+    fn size() -> usize;
+
+    #[pure]
+    fn align() -> usize;
+}
+
+macro_rules! known_size_spec {
+    ($t:ty, $size:expr, $align:expr) => {
+        #[refine_trait_spec]
+        impl KnownSize for $t {
+            #[pure]
+            #[ensures(result == $size)]
```

Comment:
This postcondition should be unnecessary, same for `align` below. (The body of a pure function becomes part of its contract…)

Reply:
Ah, I was so used to using it in extern specs that that wasn't in my mental model.